File: Python/TwoSum.py
class Solution(object):
    def twoSum(self,nums, target):
        
        """
        In this solution,
        we will just save the index and value of each element as a "value:index" pair.
        at the same time we will look for the integer which on addition makes our target int
        """

        prevMap = {}  # val -> index

        for i, n in enumerate(nums):
            diff = target - n
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i

File: cpp_leetcode/two_sum.cpp
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> umap;
        for (int i = 0; i < nums.size(); i++)
        {
            int diff = target - nums[i];
            if (umap.count(diff) != 0)
            {
                vector<int> res;
                res.push_back(i);
                res.push_back(umap[diff]);

                return res;
            }

            umap[nums[i]] = i;
        }

        vector<int> null;
        return null;
    }
};

int main()
{

    Solution sol;
    vector<int> nums = {3, 2, 4};
    int target = 6;
    vector<int> res = sol.twoSum(nums, target);

    cout << "[" << res[0] << "," << res[1] << "]" << endl;

    return 0;
}